### ISSUE TITLE
tedge-agent service lock file is not deleted when stopping the service

### ISSUE THREAD

**Issue by reubenmiller** [thin-edge/thin-edge.io#1726]
**Describe the bug**

The lock file associated with a tedge service (e.g. `tedge-agent`) is not deleted when the process is stopped (either by ctrl-c, or via the service manager). The lock file then prevents the user from starting the binary (e.g. `/usr/bin/tedge-agent`) manually running under a different user.

**To Reproduce**

The bug is simple to reproduce using the following steps:

1. Install thin-edge (>=0.9.0) on a systemd enabled Linux OS (e.g. debian)
2. Start the **tedge-agent** service
    ```sh
    sudo systemctl start tedge-agent
    ```
3. Stop the **tedge-agent** service
    ```sh
    sudo systemctl stop tedge-agent
    ```
4. Try to re-initialize **tedge**

    ```sh
    sudo tedge-agent --init
    ```

    **Result**

    The above command will be met with an error saying that the lock can not be acquired.

    ```
    The system config file '/etc/tedge/system.toml' doesn't exist. Use '/bin/systemctl' as a service manager.
    2023-02-15T12:36:24.638181449Z ERROR flockfile::unix: Another instance of tedge-agent is running.
    2023-02-15T12:36:24.638316599Z ERROR flockfile::unix: Lock file path: /run/lock/tedge-agent.lock
    Error: Couldn't acquire file lock.

    Caused by:
        Permission denied (os error 13)
    ```

The root cause seems to be the fact that the lock file still exists when the service is stopped, however since the lock file owner is set to `tedge`, when trying to run `tedge-agent` as sudo, a permission denied error occurs. The only workaround is to manually delete the lock file using `sudo rm -f /run/lock/tedge-agent.lock`.

Below shows the console output of the `tedge-agent` service and how the lock file still exists.

```sh
$ systemctl status tedge-agent
● tedge-agent.service - tedge-agent is a thin-edge.io component to support operations.
     Loaded: loaded (/lib/systemd/system/tedge-agent.service; enabled; vendor preset: enabled)
     Active: inactive (dead) since Wed 2023-02-15 14:26:24 CET; 2min 19s ago
    Process: 13170 ExecStart=/usr/bin/tedge-agent (code=killed, signal=TERM)
   Main PID: 13170 (code=killed, signal=TERM)
        CPU: 1.357s

$ ls -l /run/lock/tedge-agent.lock
-rw-r--r-- 1 tedge tedge  0 Feb 15 14:26 tedge-agent.lock
```

**Expected behavior**

If a service is stopped (and the underlying process has also been killed) then there should not be anything preventing the user from launching a new instance of 

* The user should not have to know where the stale lock file is in order to delete it
* Stopping the service via the service manager should be deemed a graceful shutdown and cleanup artifacts such as the lock file accordingly
* The user should be limited to only running `tedge-agent` (or any other binary) to a specific user (or at least they should not have to know under which previous user the process was run under)

**Screenshots**

**Environment (please complete the following information):**
 - OS [incl. version]: `Debian 12 (bookworm)`
 - Hardware [incl. revision]: `Raspberrry Pi 3 - arm7l`
 - System-Architecture [e.g. result of "uname -a"]: `Linux raspberrypi 5.15.84-v7l+ #1613 SMP Thu Jan 5 12:01:26 GMT 2023 armv7l GNU/Linux`
 - thin-edge.io version [e.g. 0.1.0]: `0.9.0`

**Additional context**

**Comment by reubenmiller** [comment]
And this "bug" / "behaviour" is the reason why we have the following snippet in the `postinst` maintainer script in the `tedge-agent` package.

```sh
# Flock file should be removed as `tedge-agent --init` must run as root
# although the flockfile is owned by `tedge`.
if [ -f "/run/lock/tedge_agent.lock" ]; then
       rm /run/lock/tedge_agent.lock
fi

if [ -f "/run/lock/tedge-agent.lock" ]; then
       rm /run/lock/tedge-agent.lock
fi
# It must be run as root to create the directory `/var/tedge`.
tedge-agent --init
```

Reference: https://github.com/thin-edge/thin-edge.io/blob/main/configuration/debian/tedge-agent/postinst#L18

**Comment by PradeepKiruvale** [comment]
If the `tedge-agent` ran as a different user then it might not have full access to the thin-edge config directory (/etc/tedge) and config files. For example, maybe can not create `.agent` file in `/etc/tedge`.

**Comment by reubenmiller** [comment]
> If the `tedge-agent` ran as a different user then it might not have full access to the thin-edge config directory (/etc/tedge) and config files. For example, maybe can not create `.agent` file in `/etc/tedge`.

I'm not saying that we should change the user it runs under, but it needs to be consistent and cleanup files accordingly for normal shutdown actions such as `systemctl stop tedge-agent`.

**Comment by didier-wenzek** [comment]
Thin-edge daemons are using [`flock`](https://linux.die.net/man/2/flock) to acquire exclusive locks. These locks are properly removed when the daemons exit or are killed, but the files supporting these locks are not removed. If created with a restrictive `umask`, these files stay with  protections that prevent other users to open them for *writing* and therefore to lock them.

The fix is simple: let any user write the lock files. This will prevent 2 concurrent processes for the same service. But not to switch from one user to another (other caveats might show then, as stressed by @PradeepKiruvale - but at least the switch will not be blocked by a previous lock).

With the following setting, I can successfully run the mapper and the agent as root or tedge:

```
$ ls -l /run/lock/
-rw-rw-rw- 1 root root  0 Feb 16 13:51 tedge-agent.lock
-rw-rw-rw- 1 root root  0 Feb 16 14:25 tedge-mapper-c8y.lock
```

Hence, if the file lock is new, one has simply to give write access to any users [on creation](https://github.com/thin-edge/thin-edge.io/blob/main/crates/common/flockfile/src/unix.rs#L59).

**Comment by reubenmiller** [comment]
After doing a deep dive together with @didier-wenzek, we have arrived at a pragmatic solution.

The proposal is to move all of the tedge/c8y plugin lock files from `/run/lock/` to `/run/lock/tedge/` which gives us more control over the "sticky" bit (e.g. the `t` option)

```sh
/run/lock/tedge/
```

* Folder should have `0o777` permissions

Some example lock files would then be:

```sh
/run/lock/tedge/tedge-agent.lock
/run/lock/tedge/tedge-mapper-c8y.lock
```

**Comment by reubenmiller** [comment]
For those interested, I found the root cause, however I still think tedge should be able to work with and without this setting, so my recommendation for a pragmatic solution remains unchanged.

It seems that a new `sysctl` setting made its way into Debian >= 11 (`fs.protected_regular=2`), which affects how the sticky bit behaves on files in a folder with it turned on. General idea seems that it helps prevent "spoofing" by restricting who can modify files created by other users (but only when the sticky bit is on for the directory).

Here is the quote taken from a [Fedora Development post](https://www.spinics.net/lists/fedora-devel/msg252452.html)

```
protected_regular:

This protection is similar to protected_fifos, but it
avoids writes to an attacker-controlled regular file, where a program
expected to create one.

When set to "0", writing to regular files is unrestricted.

When set to "1" don't allow O_CREAT open on regular files that we
don't own in world writable sticky directories, unless they are
owned by the owner of the directory.

When set to "2" it also applies to group writable sticky directories."
```
**How to turn the `fs.protected_regular` setting off (on Debian >=11 is)**

I would not recommend to turn this setting off, however the steps can be used to confirm the effect of this setting.

1. Turn the protected setting off

    ```sh
    echo "fs.protected_regular=0" | sudo tee -a /etc/sysctl.d/protect-links.conf
    ```

2. Reboot the device

    ```sh
    sudo shutdown -r now
    ```

You can also check what the `sysctl` settings are using (though this does assume you are using systemd)

```sh
/lib/systemd/systemd-sysctl --cat-config
```

**References**
* https://www.spinics.net/lists/fedora-devel/msg252452.html
* https://github.com/openzfs/zfs/issues/10742
* https://groups.google.com/g/linux.debian.bugs.dist/c/k-lbkGMPe3Y

**Comment by didier-wenzek** [comment]
> I found the root cause

Nice! Thank you to be so tenacious!

> When set to "1" don't allow O_CREAT open on regular files that we
don't own in world writable sticky directories, unless they are
owned by the owner of the directory.

Exactly, what we observed.

I agree that's better to have thin-edge working whatever this setting.

**Comment by didier-wenzek** [comment]
Fixed by https://github.com/thin-edge/thin-edge.io/pull/1740

Now, the locks for the tedge mappers and the agent are no more attached to a specific user and there is no more need to delete one of this lock if for some reason a tedge daemon has been launched under a different user.

* The locks prevent two instances of `tedge-agent` from running concurrently (as well as two concurrent version of `tedge-mapper X` for some cloud `X`). The second instance is aborted with an error "Couldn't acquire file lock".
* The locks don't block a user to run `tedge-agent`  even if the previous instance have been launched by root (and reciprocally root can launch `tedge-agent` even if a user previously run it). Similarly for `tedge-mapper`.
* A lock file set with a *previous* version of thin-edge might still have to be deleted. In that case, the error message returned by `tedge-agent` and `tedge-mapper` is related to missing authorization to open the lock file and in no case related to a concurrently running process.
* This behavior should not be affected by `fs.protected_regular=2` that prevents even the root user to write a non-owned file in `/run/lock`.

**Comment by gligorisaev** [comment]
Tested and verified that it is fixed

**Comment by reubenmiller** [comment]
Will be included officially in >=0.10.0, but it is available from the [pre-release channel](https://github.com/thin-edge/thin-edge.io/blob/main/docs/src/developer/DEBIAN_PACKAGE_HOSTING.md#pre-releases)
